src/components/Swap/Summary/index.tsx

```diff
@@ -27,59 +30,176 @@ const PriceImpactText = styled.span`
     () => Boolean(trade && ackTrade && tradeMeaningfullyDiffers(trade, ackTrade)),
     [ackTrade, trade]
   )
+  useEffect(() => {
+    if (doesTradeDiffer && reviewState === ReviewState.REVIEWING) {
+      setReviewState(ReviewState.TRADE_CHANGED)
+    } else if (!doesTradeDiffer && reviewState === ReviewState.TRADE_CHANGED) {
+      setReviewState(ReviewState.REVIEWING)
+    }
+  }, [reviewState, doesTradeDiffer])
+
+  const isApproved = useMemo(
+    () => (allowance.state === AllowanceState.REQUIRED ? allowance.isApproved : true),
+    [allowance]
+  )
 
-  const [isPending, setIsPending] = useState(false)
-  const onClick = useCallback(async () => {
-    setIsPending(true)
+  const triggerSwap = useCallback(async () => {
+    setReviewState(ReviewState.PENDING_SWAP)
     onSubmitSwapClick?.(trade)
     await onConfirm()
-    setIsPending(false)
+    setReviewState(ReviewState.REVIEWING)
   }, [onConfirm, onSubmitSwapClick, trade])
 
-  const action = useMemo((): Action | undefined => {
-    if (isPending) {
-      return {
-        message: <Trans>Confirm in your wallet</Trans>,
-        icon: Spinner,
-        onClick: () => {
-          setIsPending(false)
-        },
-      }
-    } else if (doesTradeDiffer) {
-      return {
-        color: 'accent',
-        message: <Trans>Price updated</Trans>,
-        icon: AlertTriangle,
-        onClick: () => {
-          onSwapPriceUpdateAck?.(ackTrade, trade)
-          setAckTrade(trade)
-          // Prompts parent to show speedbump if new trade has high impact
-          onAcknowledgeNewTrade()
-        },
-        children: <Trans>Accept</Trans>,
+  const prevAllowanceRef = useRef(allowance)
+  // Ensures swap isn't submitted until allowance has been properly updated post-permit2 flow
+  useEffect(() => {
+    if (prevAllowanceRef.current.state === AllowanceState.REQUIRED && allowance.state === AllowanceState.ALLOWED) {
+      // Prevents swap if trade has updated mid permit2 flow
+      if (doesTradeDiffer) setReviewState(ReviewState.TRADE_CHANGED)
+      else triggerSwap()
+    }
+    prevAllowanceRef.current = allowance
+  }, [allowance, doesTradeDiffer, triggerSwap])
+
+  const triggerPermit2Flow = useCallback(async () => {
+    if (allowance.state === AllowanceState.REQUIRED) {
+      setReviewState(ReviewState.ALLOWANCE_PENDING)
+      try {
+        await allowance.approveAndPermit?.()
+      } catch (e) {
+        console.error(e)
+        setReviewState(ReviewState.ALLOWANCE_FAILED)
       }
     }
-    return
-  }, [ackTrade, doesTradeDiffer, isPending, onAcknowledgeNewTrade, onSwapPriceUpdateAck, trade])
+  }, [allowance])
+
+  const onClick = useCallback(() => {
+    if (allowance.state === AllowanceState.REQUIRED) {
+      triggerPermit2Flow()
+      // if the user finishes permit2 allowance flow, triggerSwap() is called by useEffect above once state updates
+    } else if (allowance.state === AllowanceState.ALLOWED) {
+      triggerSwap()
+    }
```

Comment:
having all the trigger states be defined/used in separate places in this file is kind of confusing

thoughts on moving all the swap state/callbacks its own hook? it can be something like

`const { onStartSwapFlow, currentState } = useSwapCallback` that defines the useEffects here that manage allowance/swap tx state

and then this `onClick` here can just call `onStartSwapFlow`

Reply:
great point ty